torch/_functorch/aot_autograd.py

```diff
@@ -907,65 +917,72 @@ def is_fun(t):
+# (1) A FunctionalTensor(_to_functional_tensor(FakeTensor))
+# (2) A traceable tensor subclass that holds a FunctionalTensor
+# Assumption: arg promises to be the "original" tensor wrapped by f_arg
+def has_metadata_mutation(f_arg, arg, *, check_storage_mutation: bool):
+    if is_traceable_wrapper_subclass(f_arg):
+        attrs, _ = f_arg.__tensor_flatten__()
         # A tensor subclass was updated if any of its inner elements were updated
-        return any(was_metadata_updated(getattr(arg, attr), getattr(new_arg, attr)) for attr in attrs)
+        f_inner_ts = [getattr(f_arg, attr) for attr in attrs]
+        inner_ts = [getattr(arg, attr) for attr in attrs]
+        return any(has_metadata_mutation(f_inner_t, inner_t, check_storage_mutation=check_storage_mutation)
+                   for f_inner_t, inner_t in zip(f_inner_ts, inner_ts))
     else:
-        return arg is not new_arg and StorageWeakRef(arg.untyped_storage()) == StorageWeakRef(new_arg.untyped_storage())
+        if not isinstance(f_arg, torch.Tensor):
+            assert not isinstance(arg, torch.Tensor)
+            return False
+        assert isinstance(f_arg, FunctionalTensor)
+        assert isinstance(arg, FakeTensor)
+
+        arg_after = torch._from_functional_tensor(f_arg.elem)
+        # This is true if the current tensor experienced at least one set_() call
+        maybe_storage_changed = torch._functionalize_was_storage_changed(f_arg.elem)
+        # However, multiple set_() calls can cancel out. So we also check whether the
+        # storage of the tensor has changed.
+        # Note: if an input experienced two set_() calls that cancel out, **and**
+        # it experiences an data mutation, we pessimistically think that the set_()
+        # call is necessary here. We could in theory fix this, but this will
+        # hopefully never happen in user code, and is not needed for fsdp.
+        same_storages = StorageWeakRef(arg.untyped_storage()) == StorageWeakRef(arg_after.untyped_storage())
+        has_storage_metadata_mutation = maybe_storage_changed and not same_storages
+        if check_storage_mutation:
+            return has_storage_metadata_mutation
+
+        # This is true if the current tensor experienced at least one metadata mutation.
+        maybe_metadata_mutated = torch._functionalize_has_metadata_mutation(f_arg.elem)
+        # However, multi metadata mutations can cancel out.
+        # So we also check if the concrete sizes/strides on the tensor have changed.
+        same_sizes = arg.shape == arg_after.shape
+        same_strides = arg.stride() == arg_after.stride()
+        same_offsets = arg.storage_offset() == arg_after.storage_offset()
+        has_metadata_mutation_ = maybe_metadata_mutated and not (same_sizes and same_strides and same_offsets)
+        # We consider a tensor to have been metadata mutated if its storage was mutated through a set_() call.
+        return has_storage_metadata_mutation or has_metadata_mutation_
```

Comment:
And if has_storage_metadata_mutation, you could have short circuited all of the has_metadata_mutation logic